pandas/core/index.py

```diff
@@ -4393,13 +4418,23 @@ def _reference_duplicate_name(self, name):
         # count the times name equals an element in self.names.
         return sum(name == n for n in self.names) > 1
 
-    def _format_native_types(self, **kwargs):
+    def _format_native_types(self, na_rep='nan', **kwargs):
         # we go through the levels and format them
-        levels = [level._format_native_types(**kwargs)
+        levels = [level._format_native_types(na_rep=na_rep, **kwargs)
                   for level in self.levels]
         mi = MultiIndex(levels=levels, labels=self.labels, names=self.names,
                         sortorder=self.sortorder, verify_integrity=False)
-        return mi.values
+        # if any index contains only NaNs, it has collapsed into an empty
+        # Float64Index, and when the multiindex has been recomposed
+        # the NaNs have come back as NaNs, not as strings corresponding to
+        # na_rep
```

Comment:
Hu, I realize my comment is not only not-clear, but also non-accurate

Let's consider the example in my tests:
```
df = DataFrame({'a': [0, np.NaN], 'b': [0, 1], 'c': [2, 3]}).set_index(['a', 'b'])
```
The index is a multi-index, with the following levels:
```
Float64Index([0.0], dtype='float64', name=u'a')
Int64Index([0, 1], dtype='int64', name=u'b')
```
After the calls to `_format_native_types`, the variable `levels` contains:
```
['0.0']
['0', '1']
```
and afterwards `mi.values` contain:
```
[('0.0', '0'), (nan, '1')]
```
Here, `nan` is not a string `"nan"`, but `numpy.NaN` (which is printed as the string `nan`)

Then, for that reason, in the tests I've introduced, the following test fails:
```
Traceback (most recent call last):
  File "/home/nicolas/Git/pandas/pandas/tests/test_format.py", line 2965, in test_to_csv_na_rep
    self.assertEqual(df.set_index(['a', 'b']).to_csv(na_rep='_'), expected)
AssertionError: 'a,b,c\n0.0,0,2\nnan,1,3\n' != 'a,b,c\n0.0,0,2\n_,1,3\n'
```